Sprint 21

Hajime4life opened the pull request:
Рефакторинг под MVVM (async/await)

Hajime4life committed:
Рефакторинг:
- подготовка к рефакторингу на асинхронные запросы и NVVM
- перенесены модели данных из YAML в отдельный typealias
- разделен файл с константами добавлен сетевой слой
- добавлен сетевой клиент типа actor для взаимодействия с сервисами

Hajime4life committed:
Рефакторинг:
- переписаны все сервисы
- вся обработка сетевых запросов перенесена в сетевой клиент
- сервисы больше не содержат в себе инициализаторы (обязательные)
- сервисы готовы к тестам

Hajime4life committed:
Рефакторинг:
- очищены инициализаторы ViewModel’ей от создания клиента
- модель историй подписана под :Sendable
- теперь модель историй не содержит мутабельных данных
- переписана вью модель историй
-  перенесена логика фильтрации станций по РФ во viewModel
- StoryView изменен под новую модель данных
- все вью относящиеся к историям переписаны на асинхронный код

Hajime4life committed:
Рефакторинг:
- WebView перенесен в отдельный файл
- из StationSearchView вынесена логика во ViewModel

Hajime4life committed:
Рефакторинг главной ViewModel (Stations)

Hajime4life committed:
Рефакторинг:
- Подготовка к PR

NaR0RoW commented on TravelSchedule/Helper/Config/Constants.swift:
```diff
@@ -0,0 +1,8 @@
+import Foundation
+import OpenAPIRuntime
+import OpenAPIURLSession
```
Можно удалить лишние импорты

NaR0RoW commented on TravelSchedule/Helper/UI/WebView.swift:
```diff
@@ -0,0 +1,38 @@
+
+class WKWebViewController: UIViewController {
+    var url: URL?
+    private var webView: WKWebView!
```
Не стоит использовать `force` операции

NaR0RoW commented on TravelSchedule/Helper/UI/WebView.swift:
```diff
@@ -0,0 +1,38 @@
+    }
+    
+    func loadWebView() {
+        if let url = url {
+            let request = URLRequest(url: url)
+            webView.load(request)
+        }
```
Если дальнейшее выполнение кода не нужно, то лучше `guard` использовать вместо `if`

NaR0RoW commented on TravelSchedule/Helper/UI/WebView.swift:
```diff
@@ -0,0 +1,38 @@
+    }
+}
+
+class WKWebViewController: UIViewController {
```
Стоит добавить модификатор `final`, чтобы вызовы методов работали быстрее.
Но это уже со звездочкой, хотя на собеседованиях спрашивают. Можно почитать про [диспачеризацию тут](https://habr.com/ru/articles/673400/)

NaR0RoW commented on TravelSchedule/Services/CarrierService.swift:
```diff
@@ -1,26 +1,20 @@
     func getCarrierInfo(code: String) async throws -> CarrierResponse {
-        let response = try await client.getCarrierInfo(query: .init(apikey: apiKey, code: code))
-        return try response.ok.body.json
+        return try await networkClient.getCarrierInfo(code: code)
```
В однострочных выражениях `return` можно опускать

NaR0RoW commented on TravelSchedule/Services/NetworkClient.swift:
```diff
@@ -0,0 +1,117 @@
+    init(
+        apiKey: String = Constants.apiKey,
+        client: Client = Client( // не знаю где еще это делать
+            serverURL: try! Servers.Server1.url(),
```
Давай нигде не будем использовать `force` операции

Hajime4life replied:
Ну в данном случае я не знаю как избавиться от форсы

NaR0RoW commented on TravelSchedule/Services/NetworkClient.swift:
```diff
@@ -0,0 +1,117 @@
+            let responseBody = try response.ok.body.html
+            let limit = 50 * 1024 * 1024
+            let fullData = try await Data(collecting: responseBody, upTo: limit)
+            let allStations = try JSONDecoder().decode(AllStationsResponse.self, from: fullData)
```
JSONDecoder - это тяжеловесный объект. Нет смысла создавать его каждый раз внутри запроса. Лучше создать один экземпляр как свойство сервиса и переиспользовать - это экономит ресурсы и улучшает читаемость кода

NaR0RoW commented on TravelSchedule/ViewModel/CarrierViewModel.swift:
```diff
@@ -4,65 +4,65 @@ import Combine
+    func loadRaces(from: String, to: String) async {
+        guard !from.isEmpty, !to.isEmpty else {
+            await MainActor.run {
+                self.segments = []
```
Стоит поправить ошибки вида:
`Capture of 'self' with non-sendable type 'CarrierViewModel' in a '@Sendable' closure; this is an error in the Swift 6 language mode`

NaR0RoW commented on TravelSchedule/ViewModel/CarrierViewModel.swift:
```diff
@@ -4,65 +4,65 @@ import Combine
     @Published var isFilterApplied: Bool = false
     @Published var selectedTimeIntervals: Set<String> = []
     @Published var showTransferRaces: Bool? = nil
     @Published var error: NetworkError? = nil
```
`= nil` можно опустить

NaR0RoW commented on TravelSchedule/ViewModel/StationsViewModel.swift:
```diff
@@ -2,72 +2,153 @@ import SwiftUI
+        }
+        
+        do {
+            let allStationsResponse = try await stationsService.getAllStations()
```
Стоит исправить:
`Sending 'self.stationsService' risks causing data races; this is an error in the Swift 6 language mode`

NaR0RoW commented on TravelSchedule/ViewModel/StoryViewModel.swift:
```diff
@@ -1,17 +1,42 @@
+    struct StoryViewState: Identifiable, Sendable {
+        let id: Int
+        let story: Story
+        var isViewed: Bool = false
```
Лучше иммутабельные структуры делать, то есть использовать поля - константы, а не переменные. Сами поля можно менять с помощью методов, которые будут возвращать новую структуру с измененными полями

NaR0RoW commented on TravelSchedule/Views/LaunchScreenView.swift:
```diff
@@ -1,10 +1,12 @@
 struct LaunchScreenView: View {
+
     var body: some View {
         Image("launch_screen")
```
Лучше использовать `Image(.launchScreen)`, потому что это безопаснее - такие ресурсы проверяются на этапе компиляции, есть автодополнение, и меньше шансов ошибиться в названии. Строки вроде `"launch_screen"` легко написать с опечаткой, и тогда изображение просто не загрузится

NaR0RoW commented on TravelSchedule/Views/MainView.swift:
```diff
@@ -12,57 +12,65 @@ struct MainView: View {
-                    stationsViewModel.clearError()
-                    stationsViewModel.loadCities()
+                .navigationDestination(for: Screen.self, destination: destinationView)
+                .onChange(of: stationsViewModel.selectedFromStation) { from in
```
Стоит исправить:
`'onChange(of:perform:)' was deprecated in iOS 17.0: Use `onChange` with a two or zero parameter action closure instead.`

NaR0RoW commented on TravelSchedule/Views/MainView.swift:
```diff
@@ -12,57 +12,65 @@ struct MainView: View {
+                    StationSelectionView(searchTextFrom: $searchTextFrom, searchTextTo: $searchTextTo)
+                        .padding(.horizontal)
+                    
+                    if stationsViewModel.isStationsSelected {
```
Везде, где это возможно (если layout при этом не изменяется), лучше использовать `opacity` вместо `if` для сокрытия `View`. [Вот хороший пример с ссылками на `WWDC`](https://stackoverflow.com/a/57420479)

NaR0RoW commented on TravelSchedule/Views/Stories/StoriesProgressBar.swift:
```diff
@@ -1,11 +1,13 @@
 import SwiftUI
 import Combine
 
+import SwiftUI
```
Импорты дублируются

NaR0RoW commented:
Приложение работает некорректно:
1. После выбора города должны отображаться станции этого города. Сейчас отображаются все станции (видео 1).
2. Не получается выбрать станцию (видео 1).
3. Должна быть возможность выбрать любой город для поиска маршрута "Куда". Сейчас можно выбрать только город, идентичный городу из "Откуда" (видео 2).

https://github.com/user-attachments/assets/7b4d10db-e5c9-4d93-a5fc-b176234693b0

https://github.com/user-attachments/assets/98fec186-bb5f-49ef-ae5f-672efe79685b

Hajime4life committed:
Исправления:
- теперь на странице станций отображаются только те, город которых был выбран

Hajime4life committed:
Исправления:
- класс webview теперь final
- в CarrierService опущен return
- в сетевом клиенте JSONDecoder() вынесен в свойство класса
- опущена инициализация = nil там где это не надо
- название картинки загрузочного экрана указано camleCase'ом
- удален дубликат иморпта SwiftUI

Hajime4life committed:
Исправления:
- поправил yaml конфиг
- вернул обатно отображение перевозчика там где он не нужен
